Feature/rt1180 multicore support

TomasGalbickaNXP opened the pull request:
This PR adds multicore support for mimxrt1180_evk. The secondary core CM7 is copied from flash and booted from RAM.

Samples supported by this commit:

samples/drivers/mbox
samples/drivers/mbox_data
samples/subsys/ipc/ipc_service
samples/subsys/ipc/openamp

TomasGalbickaNXP committed:
boards: Add multicore support for CM7 core on RT1180

This commit adds multicore support for RT1180.
It enables the secondary core CM7 to be copied from flash
and run from RAM.

Signed-off-by: Tomas Galbicka <tomas.galbicka@nxp.com>

TomasGalbickaNXP committed:
samples: enable multicore samples for RT1180 board

This commit adds support for these multicore samples:
- samples/drivers/mbox
- samples/drivers/mbox_data
- samples/subsys/ipc/ipc_service/static_vrings
- samples/subsys/ipc/openamp

Signed-off-by: Tomas Galbicka <tomas.galbicka@nxp.com>

dleach02 commented:
@TomasGalbickaNXP can you address the CI failures?

dleach02 commented on soc/nxp/imxrt/Kconfig:
```diff
@@ -113,9 +113,10 @@ config NXP_IMX_EXTERNAL_HYPERRAM
 config SECOND_CORE_MCUX
 	bool "Dual core operation on the RT11xx series"
-	depends on SOC_SERIES_IMXRT11XX
+	depends on SOC_SERIES_IMXRT11XX || SOC_SERIES_IMXRT118X
```
How is IMXRT118x any differenct from IMXRT11XX?

TomasGalbickaNXP replied:
It is already separated into separate folder `imxrt118x`:
https://github.com/zephyrproject-rtos/zephyr/tree/main/soc/nxp/imxrt/imxrt118x
not sure why it is not in `imxrt11xx` it is then bit confusing.
https://github.com/zephyrproject-rtos/zephyr/tree/main/soc/nxp/imxrt/imxrt11xx
So I had to add `|| SOC_SERIES_IMXRT118X`

dleach02 replied:
I guess what I meant was that the SOC_SERIES_IMXRT11XX probably should include the IMXRT118x parts

TomasGalbickaNXP replied:
Yes that is true, but `SOC_SERIES_IMXRT11XX` did not include `IMXRT118x` parts from initial commit, so I would leave this change later for refactoring if possible.
I do not have knowledge why it was not structured into `imxrt11xx`.
I can only assume because of different cores CM33 and CM7, instead of CM7 and CM4.

dleach02 commented on soc/nxp/imxrt/imxrt118x/soc.c:
```diff
@@ -535,6 +547,24 @@ void soc_early_init_hook(void)
+// #if DUAL_CORE_MU_ENABLED && CONFIG_CPU_CORTEX_M4
+// 	/* Set boot flag in messaging unit to indicate boot to primary core */
+// 	MU_SetFlags(MU_BASE, BOOT_FLAG);
+// #endif
```
dead code needs to be removed

TomasGalbickaNXP replied:
Definitely, This is still Work In Progress, I need to solve why CM7 hangs on WFI instruction, then I clean up the core and create proper PR.

TomasGalbickaNXP replied:
removed

TomasGalbickaNXP committed:
drivers: nxp mbox imx mu driver fix Flags

This commit fixes flag handling in mbox driver.
There are two MU drivers mu/ and mu1/ in nxp sdk with
same API but different bit masks and register layouts.

To be compatible with both add mapping table for channel index.
This will use enum from relavant driver heder to select correct
flag mask.

Signed-off-by: Tomas Galbicka <tomas.galbicka@nxp.com>

TomasGalbickaNXP committed:
dts: soc: Add DTS MBOX and CM7 boot for NXP RT1180

This commit adds MBOX device tree entry for RT1180.

Adds functions to copy and boot CM7 core.

Adds MPU region for shared memory without caching.

Signed-off-by: Tomas Galbicka <tomas.galbicka@nxp.com>

decsny commented on drivers/mbox/mbox_nxp_imx_mu.c:
```diff
@@ -31,6 +31,76 @@ struct nxp_imx_mu_config {
 
+/*
+ * Functions to translate the channel number to Generic Interrupt Mask for mu driver.
+ * The kMU_GenInt[0..3]InterruptTrigger comes from MU driver header mu.h.
+ * There are more MU drivers implementations in one `mu` driver the
+ * kMU_GenInt0InterruptTrigger goes from 3 to 0 relative bit for the
+ * `mu1` driver kMU_GenInt0InterruptTrigger goes from 0 to 3.
+ * Same for kMU_Rx0FullFlag.
+ * Therefore use this mapping table to select correct mask based on channel index.
+ *
+ */
+static uint32_t get_gen_int_mask(const uint32_t channel)
+{
+	uint32_t mask = 0;
+
+	switch (channel) {
+	case 0:
+		mask = kMU_GenInt0InterruptTrigger;
+		break;
+
+	case 1:
+		mask = kMU_GenInt1InterruptTrigger;
+		break;
+
+	case 2:
+		mask = kMU_GenInt2InterruptTrigger;
+		break;
+
+	case 3:
+		mask = kMU_GenInt3InterruptTrigger;
+		break;
+
+	default:
+		/* Invalid channel provided */
+		assert(0);
+		break;
+	}
+
+	return mask;
+}
```
instead of making these functions, I think you can just make an array.

TomasGalbickaNXP replied:
Yes array will be better.
Indeed it is, there are two mu drivers with `almost` same API but bit different, with this we can keep them be used by this same mbox driver.

decsny reviewed (changes requested):
seems like good work overall, just finding the mailbox driver addition a bit strange

butok commented:
Please rebase.

TomasGalbickaNXP commented:
Well PR was ready to be merge for weeks.
